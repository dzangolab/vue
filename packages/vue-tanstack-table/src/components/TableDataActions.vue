<template>
  <div class="data-actions">
    <ButtonElement
      v-if="showSingleButton"
      :disabled="filteredItems[0].disabled"
      :icon-left="filteredItems[0].icon"
      :label="filteredItems[0].label"
      rounded
      variant="textOnly"
      @click="$emit('action:click')"
    />

    <Dropdown
      v-else-if="filteredItems.length > 1"
      :menu="filteredItems"
      @select="onSelectAction"
    >
      <slot name="trigger">
        <Icon icon="prime:cog" width="24" />
      </slot>
    </Dropdown>

    <ConfirmationModal
      v-if="showConfirmation"
      @on:close="showConfirmation = false"
      @on:confirm="$emit('action:select', selectedAction)"
    >
      <template
        v-if="$slots.confirmationHeader || confirmationLabel?.header"
        #header
      >
        <slot name="confirmationHeader">
          <p>{{ confirmationLabel.header }}</p>
        </slot>
      </template>

      <template v-if="$slots.confirmationBody || confirmationLabel?.body" #body>
        <slot name="confirmationBody">
          <p>{{ confirmationLabel.body }}</p>
        </slot>
      </template>

      <template
        v-if="$slots.confirmationFooter || confirmationLabel?.footer"
        #footer
      >
        <slot name="confirmationFooter">
          <p>{{ confirmationLabel.footer }}</p>
        </slot>
      </template>
    </ConfirmationModal>
  </div>
</template>

<script lang="ts">
export default {
  name: "TableDataActions",
};
</script>

<script setup lang="ts">
import { ButtonElement, ConfirmationModal, Dropdown } from "@dzangolab/vue3-ui";
import { Icon } from "@iconify/vue";
import { computed, ref } from "vue";

import type { DataActionsMenuItem } from "../types";
import type { PropType } from "vue";

const props = defineProps({
  actions: {
    type: Array as PropType<DataActionsMenuItem[]>,
    required: true,
  },
  confirmationLabel: {
    default: undefined,
    type: Object as PropType<{
      body: string;
      footer: string;
      header: string;
    }>,
  },
  data: {
    default: () => ({}),
    type: Object,
  },
});

const emit = defineEmits(["action:click", "action:select"]);

const showConfirmation = ref(false);
const selectedAction = ref<DataActionsMenuItem>();

const filteredItems = computed(() =>
  props.actions
    .filter((action) => {
      const display =
        typeof action.display === "function"
          ? action.display(props.data)
          : (action.display ?? true);
      return display;
    })
    .map((action) => ({
      ...action,
      disabled:
        typeof action.disabled === "function"
          ? action.disabled(props.data)
          : action.disabled,
    })),
);

const showSingleButton = computed(
  () => filteredItems.value.length === 1 && filteredItems.value[0].icon,
);

const onSelectAction = (action: DataActionsMenuItem) => {
  selectedAction.value = action;

  console.log(action);
  if (action?.requireConfirmationModal) {
    showConfirmation.value = true;
  } else {
    emit("action:select", selectedAction.value);
  }
};
</script>
