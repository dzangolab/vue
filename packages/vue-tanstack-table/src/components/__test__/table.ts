import {
  getCoreRowModel,
  getFilteredRowModel,
  getPaginationRowModel,
  getSortedRowModel,
  useVueTable,
} from "@tanstack/vue-table";

export const mockedTable = useVueTable({
  columns: [
    {
      accessorKey: "name",
      enableColumnFilter: true,
      header: "Name",
    },
    {
      accessorKey: "age",
      header: "Age",
      tooltip: true,
    },
    {
      accessorKey: "country",
      header: "Country",
    },
  ],
  data: [
    {
      age: 30,
      country: "USA",
      name: "John Doe",
    },
    {
      age: 25,
      country: "Canada",
      name: "Jane Smith",
    },
    {
      age: 28,
      country: "UK",
      name: "Alice Johnson",
    },
],
  getCoreRowModel: getCoreRowModel(),
  getFilteredRowModel: getFilteredRowModel(),
  getPaginationRowModel: getPaginationRowModel(),
  getSortedRowModel: getSortedRowModel(),
});