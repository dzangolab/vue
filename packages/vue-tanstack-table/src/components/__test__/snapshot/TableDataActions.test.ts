import { mount } from "@vue/test-utils";
import { describe, expect, it } from "vitest";

import { mockedTable } from "../table";
import TableDataActions from "../../TableDataActions.vue";
import { ConfirmationModal } from "@dzangolab/vue3-ui";

describe("TableDataActions", () => {
  it("matches snapshot", () => {
    const wrapper = mount(TableDataActions, {
      global: {
        mocks: {
          showConfirmation: true,
        },
        stubs: {
          ConfirmationModal: true,
          Dropdown: true,
        }
      },
      props: {
        actions: [
          {
            label: 'View',
          },
          {
            disabled: true,
            label: 'Edit',
          },
          {
            label: 'Share',
            disabled: (rowData) => rowData.id !== 11,
          },
          {
            confirmationOptions: {
              header: 'Are you sure!',
              body: 'You are going to delete this data.',
            },
            label: 'Delete',
            requireConfirmationModal: true,
          },
        ]
      },
    });

    expect(wrapper.element).toMatchSnapshot();
  });
});
