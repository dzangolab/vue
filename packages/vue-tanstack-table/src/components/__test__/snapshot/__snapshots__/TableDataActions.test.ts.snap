// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`TableDataActions > matches snapshot 1`] = `
<div
  class="data-actions"
>
  <dropdown-stub
    menu="[object Object],[object Object],[object Object],[object Object]"
  />
  
  <confirmation-modal-stub
    disablebody="false"
    disableheader="false"
    divider="false"
  />
  
</div>
`;
